src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.IO.Hashing;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.VisualStudio.Composition;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed class ExportProviderBuilder
+{
+    public record struct ExportProviderCreationArguments(
+        ImmutableArray<string> AssemblyPaths,
+        Resolver Resolver,
+        string CacheDirectory,
+        string CatalogPrefix,
+        ImmutableArray<string> ExpectedErrorParts,
+        bool PerformCleanup,
+        Action<string> LogError,
+        Action<string> LogTrace);
+
+    // For testing purposes, track the last cache write task.
+    private static Task? _cacheWriteTask;
+    private const string CatalogSuffix = ".mef-composition";
+
+    public static async Task<ExportProvider> CreateExportProviderAsync(
+        ExportProviderCreationArguments args,
+        CancellationToken cancellationToken)
+    {
+        // Clear any previous cache write task, so that it is easy to discern whether
+        // a cache write was attempted.
+        _cacheWriteTask = null;
+
+        // Get the cached MEF composition or create a new one.
+        var exportProviderFactory = await GetCompositionConfigurationAsync(args, cancellationToken).ConfigureAwait(false);
+
+        // Create an export provider, which represents a unique container of values.
+        // You can create as many of these as you want, but typically an app needs just one.
+        var exportProvider = exportProviderFactory.CreateExportProvider();
+
+        return exportProvider;
+    }
+
+    private static async Task<IExportProviderFactory> GetCompositionConfigurationAsync(
+        ExportProviderCreationArguments args,
+        CancellationToken cancellationToken)
+    {
+        var compositionCacheFile = GetCompositionCacheFilePath(args.CacheDirectory, args.CatalogPrefix, args.AssemblyPaths);
+
+        // Try to load a cached composition.
+        try
+        {
+            if (File.Exists(compositionCacheFile))
```

Comment:
what ensures that just because it exists it is ok to load and wont' be stale?  might be worth docing (here, or somewhere0.

Reply:
Added a doc comment to GetCompositionCacheFilePath